docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,66 @@
+
+Satellites will add a new field to bucket metadata called `entropy` that just contains 32 random, stable bytes.
+
+The algorithm used to derive a root key for some path in some bucket for some secret is
+
+```
+bucketEntropy = getBucketMetadata(bucketName).entropy
+bucketRootKey = argon2id(salt:hmac(hash:sha256, secret:secret, data:bucketEntropy), secret:secret)
+pathRootKey   = hmac(hash:sha256, secret:bucketRootKey, data:encryptedPath)
+```
+
+In the terminology of section 4.11 in the [whitepaper](https://storj.io/storjv3.pdf), the algorithm above describes how to derive s<sub>0</sub>. Subsequent secrets are derived using the unencrypted path components and will end up encrypted as described in the whitepaper and appended to the base encrypted path used when deriving s<sub>0</sub>.
+
+## Rationale
+
+This design accomplishes all of the requirements listed above.
+
+- Each bucket root key requires some secret that necessarily comes from the user in some way.
+- The algorithm is deterministic, so the user gets the same keys on any machine where she inputs the same secrets.
+- Root keys for a sub-path in a bucket are accomplished by providing a non-empty encrypted path when deriving the `pathRootKey`.
+- Root keys do not depend on the name of the bucket they are in: just the entropy associated with that bucket, allowing renames.
+- Because the entropy is added as the salt in the `argon2id` step, dictionary attacks on low-entropy passwords cannot be precomputed.
+- The default password can be used for any bucket and will return a distinct key, and the output of the default password derivation can be shared with others to allow them to do the same.
+
+Some other points of consideration include
+
+- Clients can ignore the bucket entropy if they wish and know they have a strong password. Indeed, they can ignore all of this and do their own key derivation.
+- The entropy is not blindly trusted from the satellite, and is mixed with the secret in an HMAC step that ensures a hostile satellite cannot predict what the salt will be.
+- HMAC provides security against forgeries without knowing the secret, which implies that sharing the output of a HMAC does not leak information about the secret, even under attacker provided data, so I believe the usage of HMAC is safe.
+- The entropy can be used for other purposes in the future if necessary.
+- Users will already lose data if they do not back up their data pointers in the satellite (because the encryption key for the data is only stored there), so adding the bucket entropy does not add a requirement that users need to back up data in a satellite, it just adds a small amount of additional information.
```

Comment:
Not sure what this means. It sounds like it's saying the encryption key for data is stored only on a satellite?

Reply:
The entropy for the bucket used to derive the salt is stored only on the satellite, just like the actual encryption key used to encrypt the data is stored only on the satellite.

Reply:
Oh, referring to the encrypted content keys. Gotcha.

Reply:
I've got the same question than @thepaul without reading this question before.
However, I'm still not understanding the response because I lack of background and probably I gonna ask a silly question, so my apologies in advance.

Does the satellite store the encryption keys for each segment?

Reply:
Yes, the pointerdb stores the encryption key and nonce used the encrypt the object's content and metadata. But this encryption key is encrypted with another encryption key that is derived from the user's password/passphrase/human-readable-key/whatever-we-call-it for the object's path. This way storing those encryption keys is secure in pointerdb.

Reply:
Gotcha, the encryption of the encryption key is what I was missing to understand that the satellite cannot read the user's data.